### ISSUE TITLE
What is the c/c++ code policy?

### ISSUE THREAD

**Issue by bmanga** [open-watcom/open-watcom-v2#323]
There are some parts of the code which would really benefit from some c++ constructs (at least overloading and maybe some small class wrappers).
Is the code meant to be kept in strict c or are we allowed to modernize/simplify it (as long as it still bootstraps)?

For example, the fact that a two-argument version of a function is called NodeArgument and the single-argument counterpart is called NodeArg is a bit non-immediate.

**Comment by jmalak** [comment]
general policy is to use C source as much as possible for best portability.
Anything what is compiled by OW tools after bootstrap can be in C++.
Any C++ code is problematic due to most C++ compilers support various language standards and not all features of standard and portability of C++ code between various platfoms and architectures is still problematic.
We have enough problems with  that our C++ compiler is old and doesn't support some current C++ feature as lazy instantiation etc.
What you give as example can be resolved by renaming function to use some naming logic or use some existing one, it is very small problem for C++ compiler sources.

**Comment by bmanga** [comment]
Perfect, thanks! :)

**Comment by jmalak** [comment]
anybody is welcome to work on OW.

**Comment by pchapin** [comment]
Moving a C project to C++ is not a change that should be undertaken lightly. I would suggest first changing all the files to have a .cpp extension as a way of "declaring" that the project is now C++ before adding even one C++ construct to an existing C program.

**Comment by sezero** [comment]
It would be sad if ow ever gets moved to c++.  (Yeah, I'm a purist.)

**Comment by pchapin** [comment]
I think one could make an argument that C++ is a more suitable language for writing compilers than C. The C language is good for systems programming and other low level work, but a compiler is not a low level application. Compiler construction benefits from high level abstractions and C++ is much better at creating such abstractions than C is.

**Comment by ArmstrongJ** [comment]
I would be against a switch to C++.  Although porting Open Watcom to another host is relatively rare, using C as opposed to C++ makes the process far easier since C compilers are far more common, especially on orphaned platforms.  Additionally, C++ is a far more complicated language, and we could lose a handful of contributors to the change if they aren't familiar with it.  Sticking with C as a rule for Open Watcom is the least of our problems.

**Comment by jmalak** [comment]
The main problem is that our C++ compiler is not compatible with any standard. If we want to switch to C++ then we must define C++ standard which can be used for development and to be able ensure OW portability to other platforms. Until OW C++ compiler will not be compatible with some most spread standard I am against any use of C++ for bootstrap code for portability reason.

**Comment by bmanga** [comment]
I was not thinking of a rewriting, but mostly of convenience wrapper classes for some of the types in the ast  nodes that need to be touched frequently (functions, variables, parameters, etc).
The problem right now it's that it's just too 'raw'. For example, accessing the type of a particular node follows completely different paths among structs of unions of structs and your way through this is by checking multiple enum fields along the way, and it's often coded as it is without any abstraction. ( To find the type of an id (!), I had to` type = expr->u.symcg.symbol->sym_type` ). 
At least gcc tries to hide most of this behind macros. ( Note that gcc also switched to allowing c++ code in their source code as well).

However, I agree with @jmalak that until we know that the subset of the language we might want to use is fully standard compliant we should not attempt anything.

**Comment by pchapin** [comment]
I agree that using C++ for bootstrap code is probably not a great idea. Although... if we knew we were always bootstrapping using Open Watcom as a build compiler, we could use C++ to whatever extent Open Watcom supported it. In effect, Open Watcom C++ would be our "standard."

I understand that we are not always bootstrapping with Open Watcom, however.

**Comment by bmanga** [comment]
I am not sure I buy that argument. Even in the case someone wants to build ow on a machine which only has a c compiler, we can start creating tags for various releases.
They can then build an older version of the compiler, and if they want the latest release they can use that to build a newer version of it.
It then all boils down to whether we consider the advantages of using some c++ constructs worth an extra 2h wait in such an uncommon, one-off scenario (I believe they are).

**Comment by pchapin** [comment]
@bmanga, wouldn't that mean we'd forever be stuck using the C++ features implemented by the last version of the Open Watcom C++ compiler that was written in C? That doesn't sound very satisfying to me.

Despite it's limitations, the Open Watcom C++ is pretty much a subset of C++98 (or C++11 if we enable that mode). Thus a fully standard C++ compiler should be able to compile anything Open Watcom can compile. So if we used C++ in our code base and limited ourselves to what Open Watcom and handle, bootstrapping that C++ with another compiler would probably be okay.

It's not completely true that Open Watcom C++ supports a proper subset of the standard, so some tweaking might be needed, but I suspect it would be minor. We'd have to keep an eye on it, though, but that's always true whenever multiple compilers are used.

@ArmstrongJ mentioned the issue of contributors: that we might lose some if we switched to C++. That could be true. On the other hand we might gain some as well. C++ people interested in the C++ compiler are more likely to want to use C++ than C for development.

**Comment by bmanga** [comment]
@pchapin We would not be forever stuck, as we may decide to do the same again (and add 2 more hours to the process). But even so, the abstraction possibilities available already with the current c++ compiler would be allow for a *huge* leap in the code readability in some parts.

As for the contributions, that is a valid point.
However, I think better abstractions and code readability in *some* parts of the code will probably have a positive effect. Just think of clang, it's written in modern and clear C++ and it proving incredibly successful at attracting new contributions, precisely because it is easy to pick up and understand.

**Comment by pchapin** [comment]
I added the C++ label to this issue because it might be of interest to those working on the C++ compiler... even if not directly about the C++ compiler (the discussion related to standards conformance is relevant, however).
